File: ios/CounterViewManager.m
#import <React/RCTViewManager.h>
#import <React/RCTUIManager.h>

@interface RCT_EXTERN_MODULE(CounterViewManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(count, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(onCountChange, RCTBubblingEventBlock)

RCT_EXTERN_METHOD(increment:(nonnull NSNumber *)reactTag)
RCT_EXTERN_METHOD(decrement:(nonnull NSNumber *)reactTag)

@end

File: ios/CounterViewManager.swift
import Foundation
import React

@objc(CounterViewManager)
class CounterViewManager: RCTViewManager {
    
    override func view() -> UIView! {
        return CounterView()
    }
    
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    @objc func increment(_ reactTag: NSNumber) {
        DispatchQueue.main.async {
            self.bridge.uiManager.addUIBlock { (uiManager, viewRegistry) in
                guard let view = viewRegistry?[reactTag] as? CounterView else {
                    return
                }
                view.increment()
            }
        }
    }
    
    @objc func decrement(_ reactTag: NSNumber) {
        DispatchQueue.main.async {
            self.bridge.uiManager.addUIBlock { (uiManager, viewRegistry) in
                guard let view = viewRegistry?[reactTag] as? CounterView else {
                    return
                }
                view.decrement()
            }
        }
    }
}

